egs2/kising/svs1/local/data_prep.py

```diff
@@ -0,0 +1,1056 @@
+    """ "
+    Get partition points [(start, end)] that detach at rests longer than the specified threshold.
+    Note: silence truncated when longer than 0.4s.
+
+    Parameters
+    -------
+    input_midi : str
+        Path to the MIDI file.
+    threshold : float
+        The threshold in beats. Rests longer than this threshold will trigger detachment.
+
+    Returns
+    -------
+    partitions : List[Tuple[float, float]]
+        The partition points. Each entry of the list is a tuple of (start, end)
+    """
+    midi_data = pretty_midi.PrettyMIDI(input_midi)
+    bpm = midi_data.estimate_tempo()
+    second_per_beat = 60 / bpm
+    thresh_in_second = threshold * second_per_beat
+
+    if len(midi_data.instruments) != 1:
+        print(f"Unexpected for kising. Multi-track found: {input_midi}!")
+        exit(1)
+
+    partitions = []  # [(start, end)]
+    instrument = midi_data.instruments[0]  # i.e., just a track
+    notes = instrument.notes
+    seg_start = max(notes[0].start - 0.2, 0.0)
+    prev_note_end = notes[0].end
+
+    for note in notes[1:]:
+        if note.start - prev_note_end >= thresh_in_second:
+            partitions.append(
+                (seg_start, min((prev_note_end + note.start) / 2, prev_note_end + 0.2))
+            )
+            seg_start = max(note.start - 0.2, (prev_note_end + note.start) / 2)
+        prev_note_end = note.end
+
+    # Note: the end of the last partition can be larger than the corresponding audio length. But does not matter
+    partitions.append((seg_start, notes[-1].end + 0.2))
+
+    return partitions
+
+
+def save_wav_segments_from_partitions(path, input_wav, partitions, songid, singer):
+    """
+    Partition wav files based on 'partitions' and save the segmented files in 'outdir'.
+    """
+
+    audio = AudioSegment.from_wav(input_wav)
+    segids = []
+    wav_scp = []
+
+    for i, (start_time, end_time) in enumerate(partitions, 1):
+        start_time_ms = int(start_time * 1000)
+        end_time_ms = int(end_time * 1000)
+
+        # Note: end_time_ms can be larger than the actual audio length. But does not matter here
+        segment = audio[start_time_ms:end_time_ms]
+        segid = f"{songid}_{i:03}_{singer}"
+        outfile = os.path.join(path, f"{segid}.wav")
+        segment.export(outfile, format="wav")
+        segids.append(segid)
+        wav_scp.append((segid, outfile))
+
+    return segids, wav_scp
+
+
+def get_info_from_partitions(
+    input_midi: str, partitions: List[Tuple[float, float]]
+) -> Tuple[
+    List[List[str]],
+    List[List[str]],
+    List[List[float]],
+    List[List[float]],
+    List[List[int]],
+]:
+    midi_data = pretty_midi.PrettyMIDI(input_midi)
+    current_partition = 0
+    text = [[]]
+    phonemes = [[]]
+    pitches = [[]]
+    note_durations = [[]]
+    phn_durations = [[]]
+    is_slur = [[]]
+
+    # Function to split Pinyin into shengmu and yunmu
+    def split_pinyin(pinyin):
```

Comment:
The dict makes the file a bit difficult to read, how about moving it to a separate file instead?